src/tools/illink/src/ILLink.Shared/DataFlow/IControlFlowGraph.cs

```diff
@@ -19,23 +20,79 @@ public enum RegionKind
 		Finally
 	}
 
+	public enum ConditionKind
+	{
+		None,
+		WhenFalse,
+		WhenTrue,
+	}
+
 	public interface IRegion<TRegion> : IEquatable<TRegion>
 	{
 		RegionKind Kind { get; }
 	}
 
+	public interface IBlock<TBlock> : IEquatable<TBlock>
+	{
+		ConditionKind ConditionKind { get; }
+	}
+
 	public interface IControlFlowGraph<TBlock, TRegion>
-		where TBlock : IEquatable<TBlock>
+		where TBlock : struct, IBlock<TBlock>
 		where TRegion : IRegion<TRegion>
 	{
 
-		public readonly struct Predecessor
+		// Represents an edge in the control flow graph, from a source to a destination basic block.
+		public readonly struct ControlFlowBranch : IEquatable<ControlFlowBranch>
 		{
-			public readonly TBlock Block;
+			public readonly TBlock Source;
+			private readonly TBlock? Destination;
+
 			public readonly ImmutableArray<TRegion> FinallyRegions;
```

Comment:
It is weird to think about "Finally reagions on an edge in the CFG" - in reality these are the finally reagions of the source block. I assume we don't model that in this class directly because we can't create the proxies here, but otherwise this should be basically `=> Source.FinallyRegions`.
Following that - I would rename this to `SourceFinallyRegions` to make it clearer.

Looking at this more - another option would be to add a `GetSourceFinallyRegions(ControlFlowBranch branch)` into the `IControlFlowGraph` interface. We model all of the other "accessors" there as well, so why do we have to make this one special. There's no state associated with the branch for the finally regions anyway, is there?

Reply:
Thanks for the explanation - might be worth a comment - I know it's "the same way as Roslyn", but Roslyn doesn't have a doc about it, so leaving something in our codebase should help.